File: 1047_Remove_All_Adjacent_Duplicates_In_String.java
class Solution {
    public String removeDuplicates(String s) {
        if (s == null || s.length() == 0)
            return "";
        Stack<Character> stack = new Stack<>();
        StringBuilder sb = new StringBuilder();
        for (char ch : s.toCharArray()) {
            if (stack.isEmpty()) {
                stack.push(ch);
            }
            else if (stack.peek() == ch) {
                stack.pop();
            }
            else {
                stack.push(ch);
            }
        }
        while (!stack.isEmpty()) {
            sb.append(stack.pop());
        }
        return sb.reverse().toString();
    }
}

/*
"abbaca"

[c a]
*/

File: Set1_1000/1047_Remove_All_Adjacent_Duplicates_In_String.py
from collections import deque
class Solution:
    def removeDuplicates(self, S: str) -> str:
        if len(S) == 0:
            return S
        stack = deque()
        for element in S:
            if not stack or stack[-1] != element:
                stack.append(element)
            elif stack and stack[-1] == element:
                stack.pop()
        return ''.join(stack)
